### DISCUSSION TITLE
SAP/node-rfc iDoc download

### DISCUSSION THREAD

**Discussion by GerritGitHub** [SAP/node-rfc#279] [General]
Hi Srdjan and other contributors to node-rfc.
I am evaluating using Node-Red to integrate an external MES system with SAP and for that purpose the use of node-rfc.
I am very impressed what I have read so-far.
I do however have a use case where SAP needs to download a process order iDOC (LOIPRO) to my MES system.
This is for the classical mechanism where the middleware would need to register itself on the SAP gateway to listen for the iDOC's. 
Is there any plans to incorporate this into node-rfc or to create a new library for this?
Regards
Gerrit Smit

**Comment by bsrdjan** [comment]
Hello Gerrit,

NodeJS RFC server, required for receiving IDOCs, is not supported yet. @samuel-c-allan started the work on this feature and the release of the contribution is not clarified yet, see https://github.com/SAP/node-rfc/issues/209#issuecomment-1404615795

We hope it can be released soon, otherwise the work must be re-done and it will take some time...
Kind regards,
Srdjan

**Comment by Ulrich-Schmidt** [comment]
Hi Gerrit,

first of all, there is no such thing as an "iDOC" (or "IDOC", as others write it)... It stands for **I**ntermediate **Doc**ument, therefore the correct spelling is "IDoc".

The next thing then is, that there is no such thing as "IDoc download". This is technically impossible. IDocs **cannot** be "_downloaded_", they can only be "_sent_" (which is equivalent to an "upload"). IDocs are always transported from the sending system to the receiving system, and this is always triggered by the sending system. A receiving system cannot trigger an IDoc transfer (which would be required for a "download").

After having clarified that ( :-) ), let's now move to the technical possibilities. As I understand, your SAP system needs to send ("push", "upload") an IDoc to the MES system. For this, the MES system would have to act as an RFC Server, but as Srdjan mentioned, the RFC Server functionality in node-rfc is not yet finished. And receiving the IDoc is only "half the battle": you also need to understand the IDoc data, and parsing an IDoc is a complicated and error-prone task. (I don't know the LOIPRO IDoc and how many nested sub-segments and complicated fields it might have, but in general except for the simple basic IDocs it's quite an effort...) For that reason, SAP has published a special "IDoc Library", which does all the work and allows easy traversing of IDoc trees and segments.

This IDoc Library is available in Java and is based on SAP JCo. So if you can use Java for integrating with the MSE system, it would make things much easier: RFC Server functionality and transactional RFC (needed for IDoc transfer) are provided by JCo, and the IDoc Lib makes understanding and processing the IDoc data a piece of cake...

A different approach might be to use the SAP Business Connector (SAP BC). This tool also has all required features for receiving and processing IDocs, and it already has a built-in transaction monitor (which JCo does not have), which is useful for logging and monitoring the IDoc status on both sides (and for finding out, what went wrong, in case something goes wrong...)
SAP BC would be set up as RFC Server (SM59), receive the IDoc, parse it and map it into the required format that is understood by the MES system, and then push the data into MES using any protocol of your choice (like HTTP, FTP, Email, flatfile interface, JDBC, XML, SOAP, OData, JSon or whatever...)

See https://support.sap.com/sbc-download for downloading the SAP BC. If you are interested, I can provide more details/documentation here.

Best Regards, Ulrich

**Reply by bsrdjan** [reply]
Check this comment as well: https://github.com/SAP/PyRFC/discussions/340#discussioncomment-7326815
